File: hex_terrain_editor/editor/pen_tool.py
"""Pen tool for freehand drawing roads and rivers."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..model.path_feature import PathFeature
from ..model.terrain_types import PathType
from ..utils.color_palettes import ROAD_STYLES, RIVER_STYLES
from ..utils.curve_simplify import rdp_simplify

if TYPE_CHECKING:
    from ..model.project import Project
    from .tool_manager import ToolManager


class PenTool:
    """Freehand draw roads/rivers with curve simplification."""

    def __init__(
        self,
        project: "Project",
        tool_manager: "ToolManager",
        on_path_added: Any = None,
    ) -> None:
        self.project = project
        self.tool_manager = tool_manager
        self.on_path_added = on_path_added

        self.path_type: PathType = PathType.ROAD
        self.raw_points: list[tuple[float, float]] = []
        self._drawing = False

    def activate(self) -> None:
        pass

    def deactivate(self) -> None:
        self._drawing = False
        self.raw_points.clear()

    def on_press(self, world_x: float, world_y: float) -> None:
        self._drawing = True
        self.raw_points = [(world_x, world_y)]

    def on_move(self, world_x: float, world_y: float) -> None:
        if self._drawing:
            self.raw_points.append((world_x, world_y))

    def on_release(self, world_x: float, world_y: float) -> None:
        if not self._drawing:
            return
        self._drawing = False
        self.raw_points.append((world_x, world_y))

        if len(self.raw_points) < 2:
            self.raw_points.clear()
            return

        # Simplify the polyline
        simplified = rdp_simplify(self.raw_points, epsilon=5.0)

        # Get width from style
        ft = self.path_type.value
        if ft in ROAD_STYLES:
            width = ROAD_STYLES[ft]["width"]
        elif ft in RIVER_STYLES:
            width = RIVER_STYLES[ft]["width"]
        else:
            width = 2.0

        path = PathFeature(
            feature_type=self.path_type,
            control_points=simplified,
            width=width,
        )

        self.project.add_path(path)

        self.tool_manager.push_undo(
            f"Draw {self.path_type.value}",
            {"action": "add_path", "path_id": path.id},
            {"action": "add_path", "path": path.to_dict()},
        )

        if self.on_path_added:
            self.on_path_added(path)

        self.raw_points.clear()

    @property
    def preview_points(self) -> list[tuple[float, float]]:
        """Return current raw points for preview rendering."""
        return list(self.raw_points)

File: src/editor/pen_tool.rs
use crate::model::path_feature::PathFeature;
use crate::model::terrain_types::PathType;
use super::curve_simplify::rdp_simplify;

pub struct PenTool {
    pub path_type: PathType,
    raw_points: Vec<[f32; 2]>,
    drawing: bool,
}

impl PenTool {
    pub fn new() -> Self {
        Self {
            path_type: PathType::Road,
            raw_points: Vec::new(),
            drawing: false,
        }
    }

    pub fn on_press(&mut self, world_x: f32, world_y: f32) {
        self.drawing = true;
        self.raw_points.clear();
        self.raw_points.push([world_x, world_y]);
    }

    pub fn on_move(&mut self, world_x: f32, world_y: f32) {
        if self.drawing {
            self.raw_points.push([world_x, world_y]);
        }
    }

    /// Finish drawing and return the simplified path feature, if any.
    pub fn on_release(&mut self) -> Option<PathFeature> {
        if !self.drawing {
            return None;
        }
        self.drawing = false;

        if self.raw_points.len() < 2 {
            self.raw_points.clear();
            return None;
        }

        let epsilon = 2.0;
        let simplified = rdp_simplify(&self.raw_points, epsilon);
        self.raw_points.clear();

        if simplified.len() < 2 {
            return None;
        }

        let mut feature = PathFeature::new(self.path_type);
        feature.control_points = simplified;
        Some(feature)
    }

    pub fn is_drawing(&self) -> bool {
        self.drawing
    }

    /// Return the raw points collected so far (for live preview).
    pub fn preview_points(&self) -> &[[f32; 2]] {
        &self.raw_points
    }
}

impl Default for PenTool {
    fn default() -> Self {
        Self::new()
    }
}
